File: harness/platform_core_contract/references.py
"""Actor, entity, record, and product scope references."""

from __future__ import annotations

import re
from typing import Any

from . import identity
from .codec import (ContractError, exact_object, validate_hash,
                    validate_schema_name, validate_text)
from .constants import REFERENCE_MISMATCH, SCOPE_FIELDS

_ENTITY_FIELDS = {"entity_id", "entity_type"}
_ACTOR_FIELDS = {"actor_id", "actor_type"}
_RECORD_FIELDS = {"record_id", "record_sha256", "record_type"}
_ACTOR_TYPES = {"agent", "harness", "human", "service", "system"}


def validate_entity_ref(value: Any, label: str) -> dict[str, Any]:
    """Return one exact typed EntityRef."""
    node = exact_object(value, _ENTITY_FIELDS, label)
    identity.validate_entity_id(node["entity_type"], node["entity_id"], label)
    return node


def validate_actor_ref(value: Any) -> dict[str, Any]:
    """Return one exact caller-declared ActorRef."""
    node = exact_object(value, _ACTOR_FIELDS, "actor_ref")
    identity.validate_typed_id(node["actor_id"], "acr", "actor_ref.actor_id")
    if type(node["actor_type"]) is not str or node["actor_type"] not in _ACTOR_TYPES:
        raise ContractError("actor_ref.actor_type is unsupported")
    return node


def validate_record_ref(value: Any, label: str) -> dict[str, Any]:
    """Return one exact unresolved RecordRef."""
    node = exact_object(value, _RECORD_FIELDS, label)
    record_id = validate_text(node["record_id"], f"{label}.record_id", 160)
    if re.fullmatch(r"[a-z][a-z0-9._:/-]*", record_id) is None:
        raise ContractError(f"{label}.record_id has invalid opaque identifier text")
    validate_hash(node["record_sha256"], f"{label}.record_sha256")
    validate_schema_name(node["record_type"], f"{label}.record_type")
    return node


def validate_scope(value: Any) -> dict[str, Any]:
    """Return one exact contiguous product ScopeRef."""
    node = exact_object(value, SCOPE_FIELDS, "scope_ref")
    validate_scope_values(node)
    validate_scope_references(node)
    return node


def validate_scope_values(node: dict[str, Any]) -> None:
    """Validate only ScopeRef identifiers and field vocabulary."""
    identity.validate_typed_id(node["space_id"], "spc", "scope_ref.space_id")
    prefixes = {
        "action_id": "act", "attempt_id": "atm", "change_id": "chg",
        "objective_id": "obj", "project_id": "prj",
        "project_snapshot_id": "psn", "session_id": "ses", "turn_id": "trn",
        "work_graph_id": "wgr", "work_item_id": "wki",
    }
    for field, prefix in prefixes.items():
        if node[field] is not None:
            identity.validate_typed_id(node[field], prefix, f"scope_ref.{field}")


def validate_scope_references(node: dict[str, Any]) -> None:
    """Validate parent/child bindings after all field values are valid."""
    _validate_ancestry(node)


def scope_contains(scope: dict[str, Any], reference: dict[str, Any]) -> bool:
    """Whether one scoped product identity equals the supplied EntityRef."""
    field_by_type = {
        "action": "action_id", "attempt": "attempt_id", "change": "change_id",
        "objective": "objective_id", "project": "project_id",
        "project_snapshot": "project_snapshot_id", "session": "session_id",
        "turn": "turn_id", "work_graph": "work_graph_id", "work_item": "work_item_id",
    }
    if reference["entity_type"] == "space":
        return reference["entity_id"] == scope["space_id"]
    field = field_by_type.get(reference["entity_type"])
    return field is not None and reference["entity_id"] == scope[field]


def _validate_ancestry(scope: dict[str, Any]) -> None:
    links = (
        ("project_snapshot_id", "project_id"), ("change_id", "objective_id"),
        ("work_graph_id", "change_id"), ("work_item_id", "work_graph_id"),
        ("attempt_id", "work_item_id"), ("session_id", "attempt_id"),
        ("turn_id", "session_id"), ("action_id", "turn_id"),
    )
    for child, parent in links:
        if scope[child] is not None and scope[parent] is None:
            raise ContractError(f"scope_ref.{child} requires {parent}",
                                REFERENCE_MISMATCH)

File: forge-runtime/crates/domain/src/platform_core_contract/references.rs
use super::{
    ActorRef, ActorType, EntityRef, EntityType, PlatformCoreContractError, RecordRef,
    RejectionCode, ScopeRef, identity, invalid, reject, wire,
};

/// Validates one supplied typed entity reference without resolving it.
///
/// # Errors
/// Returns an error when the entity type or Platform ID is invalid.
pub(crate) fn validate_entity_ref(
    value: &EntityRef,
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    identity::validate_entity_id(&value.entity_type, &value.entity_id, label)
}

pub(super) fn validate_actor_ref(value: &ActorRef) -> Result<(), PlatformCoreContractError> {
    identity::validate_typed_id(&value.actor_id, "acr", "actor_ref.actor_id")?;
    if matches!(&value.actor_type, ActorType::Unknown(_)) {
        return Err(reject(
            RejectionCode::ValueInvalid,
            "actor_ref.actor_type is unsupported",
        ));
    }
    Ok(())
}

/// Validates one supplied opaque record reference without resolving it.
///
/// # Errors
/// Returns an error when its identifier, digest, or record type is invalid.
pub(crate) fn validate_record_ref(
    value: &RecordRef,
    label: &str,
) -> Result<(), PlatformCoreContractError> {
    validate_record_id(&value.record_id, &format!("{label}.record_id"))?;
    wire::validate_hash(&value.record_sha256, &format!("{label}.record_sha256"))?;
    wire::validate_schema_name(&value.record_type, &format!("{label}.record_type"))
}

/// Validates supplied scope identities and their structural ancestry.
///
/// # Errors
/// Returns an error when an ID is invalid or a child lacks its required parent.
pub(crate) fn validate_scope_ref(value: &ScopeRef) -> Result<(), PlatformCoreContractError> {
    validate_scope_values(value)?;
    validate_scope_references(value)
}

pub(super) fn validate_scope_values(value: &ScopeRef) -> Result<(), PlatformCoreContractError> {
    identity::validate_typed_id(&value.space_id, "spc", "scope_ref.space_id")?;
    let checks = [
        (&value.action_id, "act", "action_id"),
        (&value.attempt_id, "atm", "attempt_id"),
        (&value.change_id, "chg", "change_id"),
        (&value.objective_id, "obj", "objective_id"),
        (&value.project_id, "prj", "project_id"),
        (&value.project_snapshot_id, "psn", "project_snapshot_id"),
        (&value.session_id, "ses", "session_id"),
        (&value.turn_id, "trn", "turn_id"),
        (&value.work_graph_id, "wgr", "work_graph_id"),
        (&value.work_item_id, "wki", "work_item_id"),
    ];
    for (candidate, prefix, label) in checks {
        if let Some(candidate) = candidate {
            identity::validate_typed_id(candidate, prefix, &format!("scope_ref.{label}"))?;
        }
    }
    Ok(())
}

pub(super) fn validate_scope_references(value: &ScopeRef) -> Result<(), PlatformCoreContractError> {
    validate_scope_ancestry(value)
}

pub(super) fn scope_contains(scope: &ScopeRef, reference: &EntityRef) -> bool {
    match &reference.entity_type {
        EntityType::Space => reference.entity_id == scope.space_id,
        EntityType::Project => matches_scope(&reference.entity_id, scope.project_id.as_deref()),
        EntityType::ProjectSnapshot => {
            matches_scope(&reference.entity_id, scope.project_snapshot_id.as_deref())
        }
        EntityType::Objective => matches_scope(&reference.entity_id, scope.objective_id.as_deref()),
        EntityType::Change => matches_scope(&reference.entity_id, scope.change_id.as_deref()),
        EntityType::WorkGraph => {
            matches_scope(&reference.entity_id, scope.work_graph_id.as_deref())
        }
        EntityType::WorkItem => matches_scope(&reference.entity_id, scope.work_item_id.as_deref()),
        EntityType::Attempt => matches_scope(&reference.entity_id, scope.attempt_id.as_deref()),
        EntityType::Session => matches_scope(&reference.entity_id, scope.session_id.as_deref()),
        EntityType::Turn => matches_scope(&reference.entity_id, scope.turn_id.as_deref()),
        EntityType::Action => matches_scope(&reference.entity_id, scope.action_id.as_deref()),
        EntityType::Actor | EntityType::Artifact | EntityType::Receipt | EntityType::Unknown(_) => {
            false
        }
    }
}

fn validate_record_id(value: &str, label: &str) -> Result<(), PlatformCoreContractError> {
    wire::validate_text(value, label, 160, true)?;
    let valid = value.bytes().enumerate().all(|(index, byte)| {
        byte.is_ascii_lowercase()
            || index > 0
                && (byte.is_ascii_digit() || matches!(byte, b'.' | b'_' | b':' | b'/' | b'-'))
    });
    if valid {
        Ok(())
    } else {
        Err(invalid(format!(
            "{label} has invalid opaque record identifier text"
        )))
    }
}

fn validate_scope_ancestry(value: &ScopeRef) -> Result<(), PlatformCoreContractError> {
    require_parent(
        value.project_snapshot_id.is_some(),
        value.project_id.is_some(),
        "project_snapshot_id requires project_id",
    )?;
    require_parent(
        value.change_id.is_some(),
        value.objective_id.is_some(),
        "change_id requires objective_id",
    )?;
    require_parent(
        value.work_graph_id.is_some(),
        value.change_id.is_some(),
        "work_graph_id requires change_id",
    )?;
    require_parent(
        value.work_item_id.is_some(),
        value.work_graph_id.is_some(),
        "work_item_id requires work_graph_id",
    )?;
    require_parent(
        value.attempt_id.is_some(),
        value.work_item_id.is_some(),
        "attempt_id requires work_item_id",
    )?;
    require_parent(
        value.session_id.is_some(),
        value.attempt_id.is_some(),
        "session_id requires attempt_id",
    )?;
    require_parent(
        value.turn_id.is_some(),
        value.session_id.is_some(),
        "turn_id requires session_id",
    )?;
    require_parent(
        value.action_id.is_some(),
        value.turn_id.is_some(),
        "action_id requires turn_id",
    )?;
    Ok(())
}

fn require_parent(child: bool, parent: bool, label: &str) -> Result<(), PlatformCoreContractError> {
    if child && !parent {
        Err(reject(
            RejectionCode::ReferenceMismatch,
            format!("scope_ref.{label}"),
        ))
    } else {
        Ok(())
    }
}

fn matches_scope(expected: &str, candidate: Option<&str>) -> bool {
    candidate == Some(expected)
}
